pkg/status/combinedstatus-resolver.go

```diff
@@ -0,0 +1,470 @@
+			return resolution.generateCombinedStatus(bindingName, objectIdentifier)
+		}
+	}
+
+	return nil
+}
+
+// CompareCombinedStatus compares the given CombinedStatus object with the
+// one associated with the given binding name and workload object identifier.
+// True is returned in case of a match, false otherwise.
+func (c *combinedStatusResolver) CompareCombinedStatus(bindingName string,
+	objectIdentifier util.ObjectIdentifier, combinedStatus *v1alpha1.CombinedStatus) bool {
+	c.RLock()
+	defer c.RUnlock()
+
+	if resolutions, exists := c.bindingNameToResolutions[bindingName]; exists {
+		if resolution, exists := resolutions[objectIdentifier]; exists {
+			return resolution.compareCombinedStatus(combinedStatus)
+		}
+	}
+
+	return false
+}
+
+// NoteBindingResolution notes a binding resolution for status collection.
+//
+// 1. If `deleted` is true, the associated combinedstatus resolutions are
+// removed from memory. The same is done if a resolution no longer requires
+// status collection.
+//
+// 2. Excessive combinedstatus resolutions are removed if they are no
+// longer associated with the binding.
+//
+// 3. For every workload object associated with one or more
+// statuscollectors, a combinedstatus resolution is created/updated.
+// The update may involve adding or removing statuscollectors, and changing
+// the set of destinations associated with the binding.
+//
+// The function uses the statuscollector lister and the cached workstatus
+// objects to update internal state. The given array is assumed to contain
+// all objects that may be relevant to the binding.
+//
+// The returned set contains the identifiers of combinedstatus objects
+// that should be queued for syncing.
+func (c *combinedStatusResolver) NoteBindingResolution(bindingName string, bindingResolution binding.Resolution,
+	deleted bool, workStatuses []*workStatus,
+	statusCollectorLister controllisters.StatusCollectorLister) sets.Set[util.ObjectIdentifier] {
+	c.Lock()
+	defer c.Unlock()
+
+	combinedStatusIdentifiersToQueue := sets.New[util.ObjectIdentifier]()
+	workloadIdentifiersToEvaluate := sets.New[util.ObjectIdentifier]()
+	// destinations set to collect from, shared for all workload objects
+	collectionDestinations := sets.New(abstract.SliceMap(bindingResolution.Destinations,
+		func(destination v1alpha1.Destination) string { return destination.ClusterId })...)
+
+	// (1)
+	if deleted {
+		// for every combinedstatus resolution - queue its object identifier for syncing
+		// and remove the resolution from memory
+		for workloadObjectIdentifier, _ := range c.bindingNameToResolutions[bindingName] {
+			combinedStatusIdentifiersToQueue.Insert(getCombinedStatusIdentifier(bindingName, workloadObjectIdentifier))
+		}
+
+		delete(c.bindingNameToResolutions, bindingName)
+		return combinedStatusIdentifiersToQueue
+	}
+
+	// if the binding resolution is not yet noted - create a new entry
+	objectIdentifierToResolutions, exists := c.bindingNameToResolutions[bindingName]
```

Comment:
FYI, I would name it `objectIdentifierToResolution` (singular) because each OID is mapped to just one resolution.